cmd/metal-api/internal/metal/network.go

```diff
@@ -18,6 +45,103 @@ type Nic struct {
+	State      *NicState  `rethinkdb:"state" json:"state"`
+}
+
+// NicState represents the desired and actual state of a network interface
+// controller (NIC). The Desired field indicates the intended state of the
+// NIC, while Actual indicates its current operational state. The Desired
+// state will be removed when the actual state is equal to the desired state.
+type NicState struct {
+	Desired *SwitchPortStatus `rethinkdb:"desired" json:"desired"`
+	Actual  SwitchPortStatus  `rethinkdb:"actual" json:"actual"`
+}
+
+// SetState updates the NicState with the given SwitchPortStatus. It returns
+// a new NicState and a bool indicating if the state was changed.
+//
+// If the given status matches the current Actual state, it checks if Desired
+// is set and matches too. If so, Desired is set to nil since the desired
+// state has been reached.
+//
+// If the given status differs from the current Actual state, Desired is left
+// unchanged if it differes from the new state so the desired state is still tracked.
+// The Actual state is updated to the given status.
+//
+// This allows tracking both the desired and actual states, while clearing
+// Desired once the desired state is achieved.
+func (ns *NicState) SetState(s SwitchPortStatus) (NicState, bool) {
+	if ns == nil {
+		return NicState{
+			Actual:  s,
+			Desired: nil,
+		}, true
+	}
+	if ns.Actual == s {
+		if ns.Desired != nil {
+			if *ns.Desired == s {
+				// we now have the desired state, so set the desired state to nil
+				return NicState{
+					Actual:  s,
+					Desired: nil,
+				}, true
+			} else {
+				// we already have the reported state, but the desired one is different
+				// so nothing changed
+				return *ns, false
+			}
+		}
+		// nothing changed
+		return *ns, false
+	}
+	// we got another state as we had before
+	if ns.Desired != nil {
+		if *ns.Desired == s {
+			// we now have the desired state, so set the desired state to nil
+			return NicState{
+				Actual:  s,
+				Desired: nil,
+			}, true
+		} else {
+			// we already have the reported state, but the desired one is different
+			// so nothing changed
+			return NicState{
+				Actual:  s,
+				Desired: ns.Desired,
+			}, true
+		}
+	}
+	return NicState{
+		Actual:  s,
+		Desired: nil,
+	}, true
+}
+
+// WantState sets the desired state for the NIC. It returns a new NicState
+// struct with the desired state set and a bool indicating if the state changed.
+// If the current state already matches the desired state, it returns a state
+// with a cleared desired field.
+func (ns *NicState) WantState(s SwitchPortStatus) (NicState, bool) {
+	if ns == nil {
+		return NicState{
+			Actual:  SwitchPortStatusUnknown,
+			Desired: &s,
+		}, true
+	}
+	if ns.Actual == s {
+		// we want a state we already have
+		if ns.Desired != nil {
+			return NicState{
+				Actual:  s,
+				Desired: nil,
+			}, true
+		}
+		return *ns, false
+	}
+	return NicState{
+		Actual:  ns.Actual,
+		Desired: &s,
+	}, true
```

Comment:
Hope I am not mistaken, but I think it's not covering all cases, for example let's assume the following nic state:

``` 
{"actual": "UP", "desired:" "DOWN"}
```

The user then sends `DOWN` a second time, which will lead to the right return but it will indicate a changed state with `true` but it actually did not change anything.